src/util/locktime.rs

```diff
@@ -0,0 +1,233 @@
+    /// Constructs a `LockTime` from an nLockTime value or the argument to OP_CHEKCLOCKTIMEVERIFY.
+    fn from(n: u32) -> Self {
+        if encodes_height(n) {
+            Self::Blocks(Height::new(n).expect("n is valid"))
+        } else {
+            Self::Seconds(Time::new(n).expect("n is valid"))
+        }
+    }
+}
+
+impl From<Height> for LockTime {
+    fn from(h: Height) -> Self {
+        LockTime::Blocks(h)
+    }
+}
+
+impl From<Time> for LockTime {
+    fn from(h: Time) -> Self {
+        LockTime::Seconds(h)
+    }
+}
+
+/// An absolute block height.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Height(u32);
+
+impl Height {
+    /// Constructs a new block height.
+    ///
+    /// # Errors
+    /// If `n` does not encode a block height value (see [`encodes_height`]).
+    pub fn new(n: u32) -> Result<Self, Error> {
+        if encodes_height(n) {
+            Ok(Self(n))
+        } else {
+            Err(Error::InvalidHeight)
+        }
+    }
+
+    /// Converts `Height` to its inner `u32` value.
+    pub fn to_u32(&self) -> u32 {
+        self.0
+    }
+}
+
+/// A UNIX timestamp, seconds since epoch.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Time(u32);
+
+impl Time {
+    /// Constructs a new block time.
+    ///
+    /// # Errors
+    /// If `n` does not encode a UNIX time stamp (see [`encodes_time`]).
+    pub fn new(n: u32) -> Result<Self, Error> {
+        if encodes_time(n) {
+            Ok(Self(n))
+        } else {
+            Err(Error::InvalidTime)
+        }
+    }
+
+    /// Converts `Time` to its inner `u32` value.
+    pub fn to_u32(&self) -> u32 {
+        self.0
+    }
+}
+
+/// Returns true if `n` encodes a block height i.e., is < 500,000,000.
+pub fn encodes_height(n: u32) -> bool {
+    n < LOCKTIME_THRESHOLD
+}
+
+/// Returns true if `n` encodes a UNIX timestamp i.e., is >= 500,000,000.
+pub fn encodes_time(n: u32) -> bool {
+    n >= LOCKTIME_THRESHOLD
+}
+
+/// Returns true if the timelock constraint of `n OP_CHECKLOCKTIMEVERIFY` and nLockTime is
+/// satisfied.
+///
+/// # Errors
+///
+/// If `n` and `n_lock_time` are different units i.e., one represents a block height and the other
+/// represents a block time.
+pub fn is_satisfied(n: LockTime, n_lock_time: LockTime) -> Result<bool, Error> {
+    if !n.is_same_unit(n_lock_time) {
+        return Err(Error::InvalidComparison);
+    }
+    Ok(n_lock_time.to_u32() >= n.to_u32())
+}
+
+impl fmt::Display for LockTime {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            LockTime::Blocks(ref t) => write!(f, "{}", t.0),
+            LockTime::Seconds(ref t) => write!(f, "{}", t.0),
```

Comment:
Initially I added some text to the integer but stuff broke so I did it like this without much further thought. IIRC we use `Display` to roundtrip the descriptor to/from a string, which expects the locktimes to be a plain integer. I'll dig a little deeper.

Reply:
Turns out I was wrong, nothing in `rust-miniscript` breaks if we use, for example
```
            LockTime::Blocks(ref t) => write!(f, "block-height {}", t.0),
            LockTime::Seconds(ref t) => write!(f, "block-time {} (seconds since epoch)", t.0),
```

Reply:
Sure, but it would be nice if Miniscript could just use `Display` like it does for keys and hashes.

Reply:
I went with defaulting to just printing the u32 and adding additional context when printed with `{:#}`.